### PR TITLE
[W4Mconcat] First version of a new tool to concatenate W4M metadata tables

### PR DESCRIPTION
**Pull request by hanourine**
Hello, I am currently in intership at PFEM-MetaboHub and I have developed a tool to merge Metadata tables according to W4M 3-tables format. I have tested my tool on planemo with the latest Galaxy VM and all works fine. I am looking forward to your review, in particular regarding Galaxy best practices for galaxy wrapper. 
Thank you in advance.
   Hanane Nourine

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-metabolomics/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/w4mconcatenate/concat.xml [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  
+  
+  
+   </test>
+  
+  </tests>
+  
+  
+  
+
```
Comment (lecorguille) [inline tools/w4mconcatenate/concat.xml:132]: Can you reduce the number of empty lines?

File: tools/w4mconcatenate/concatenation_wrapper.R [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (lecorguille) [inline tools/w4mconcatenate/concatenation_wrapper.R:88]: Remove extra lines

File: tools/w4mconcatenate/fonctions_auxiliaires.R [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      non_identical_columns_v <- c(non_identical_columns_v, col_name)
+    
+      #otherwise store the columns where the values are not the same in non_identical_columns
+    } else{ 
+  
+      M2 <- M2[,-which(colnames(M2) == col_name)] 
+      
+    }
+    
+    }
```
Comment (lecorguille) [inline tools/w4mconcatenate/fonctions_auxiliaires.R:73]: Indentation

File: tools/w4mconcatenate/fonctions_auxiliaires.R [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    
+    result <- list(M1 = M1, M2 = M2)
+    
+    return(result)
+    
+    
+  }
+   
+
+    else{
```
Comment (lecorguille) [inline tools/w4mconcatenate/fonctions_auxiliaires.R:86]: Same identation

**Review by lecorguille — CHANGES_REQUESTED** [review]

File: tools/w4mconcatenate/.sheld.yml.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+categories: [Metabolomics]
```
Comment (lecorguille) [inline tools/w4mconcatenate/.sheld.yml.txt:1]: You have to remove the `.txt` extension

**Review by lecorguille — CHANGES_REQUESTED** [review]

File: tools/w4mconcatenate/.sheld.yml.txt (hunk 5 above)
Comment (hanourine) [reply tools/w4mconcatenate/.sheld.yml.txt]: Corrected in 791ad01

File: tools/w4mconcatenate/concat.xml [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  
+   <param name="concatenation" type="select" display="radio" label="Type of concatenation">
+    <option value="unique">Unique</option>
+    <option value="intersection">Intersection</option>
+    <option value="union">Union</option>
+   </param>
+  
+   <conditional name="choice">
+    <param name="choice_keep" type="select" display="radio" label="Keep all or just one">
+     <option value="oui">Keep all</option>
```
[the remark this answers is not in the document]
Comment (hanourine) [reply tools/w4mconcatenate/concat.xml]: Corrected in 9d3fe1a

File: tools/w4mconcatenate/fonctions_auxiliaires.R (hunk 4 above)
Comment (hanourine) [reply tools/w4mconcatenate/fonctions_auxiliaires.R]: Corrected in 403025a

File: tools/w4mconcatenate/fonctions_auxiliaires.R (hunk 3 above)
Comment (hanourine) [reply tools/w4mconcatenate/fonctions_auxiliaires.R]: Corrected in 403025a

File: tools/w4mconcatenate/concatenation_wrapper.R (hunk 2 above)
Comment (hanourine) [reply tools/w4mconcatenate/concatenation_wrapper.R]: Corrected in 403025a

File: tools/w4mconcatenate/concat.xml (hunk 1 above)
Comment (hanourine) [reply tools/w4mconcatenate/concat.xml]: Corrected in 403025a

**Comment by hanourine** [comment]
Hi everyone,
Thank you @lecorguille for the review. I tried to answer all the points along with minor cosmetics enhancement.
Is the current version ok?

File: tools/w4mconcatenate/concat.xml [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    #if str($choice.choice_keep) == 'no':
+    keep "$choice.keep"
+    #end if
+    
+    dataMatrix_1_out "$dataMatrix_1_out"
+    
+    dataMatrix_2_out "$dataMatrix_2_out"
+    
+    metadata_out "$metadata_out"
+
```
Comment (bgruening) [inline tools/w4mconcatenate/concat.xml:42]:
```suggestion
  <![CDATA[
    
    Rscript '$__tool_directory__/concatenation_wrapper.R'
    
    dataMatrix_1 '$dataMatrix_1'
    
    dataMatrix_2 '$dataMatrix_2'
    
    metadata_1 '$metadata_1'
    
    metadata_2 '$metadata_2'
    
    type '$type'
    
    tab1 '$tab1'
    
    tab2 '$tab2'
    
    concatenation '$concatenation'
    
    choice_keep '$choice.choice_keep'
    #if str($choice.choice_keep) == 'yes':
        keep 0
    #end if
    #if str($choice.choice_keep) == 'no':
        keep "$choice.keep"
    #end if
    
    dataMatrix_1_out '$dataMatrix_1_out'
    dataMatrix_2_out '$dataMatrix_2_out'
    metadata_out '$metadata_out'
    
```

**Comment by bgruening** [comment]
@hanourine thanks a lot!

Can you maybe look at this tool: https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/iuc/table_compute/table_compute/1.2.4+galaxy0

I know its a bit complex, but I assume it can handle your use case and many more. Maybe that is useful?

**Comment by hanourine** [comment]
> @hanourine thanks a lot!
> 
> Can you maybe look at this tool: https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/iuc/table_compute/table_compute/1.2.4+galaxy0
> 
> I know its a bit complex, but I assume it can handle your use case and many more. Maybe that is useful?

Thank you for your comment. 
Our module handles a set of 4 files, whereas the Table Compute module would require us to use the module 3 times, increasing the risk of error. In addition, the W4Mconcat options don't behave as expected and can't be configured as in W4Mconcat. For example:
The error is green but false in the case of input the wrong files that are not supposed to match rather than an error to the user, and on the other hand the case of use with different individuals.
In the case of common columns between the two tables, Table compute aggregates the columns instead of being able to choose between discerning one of the two columns or keeping both.
Thus, the use of Table compute requires a great deal of manipulation to obtain results where the purpose of W4Mconcat is precisely to allow a simple way and limiting false manipulation in order to obtain a consistent set of data.

**Comment by lecorguille** [comment]
lintr 😋

**Review by lecorguille — APPROVED** [review]

**Pull request merged**